### PR TITLE
ed25519_verify: Support using dalek for historical blocks

### PR DESCRIPTION
**Pull request by bkchr**
The switch from `ed25519-dalek` to `ed25519-zebra` was actually a breaking change. `ed25519-zebra` is more permissive. To support historical blocks when syncing a chain this pull request introduces an externalities extension `UseDalekExt`. This extension is just used as a signaling mechanism to `ed25519_verify` to use `ed25519-dalek` when it is present. Together with `ExtensionBeforeBlock` it can be used to setup a node in way to sync historical blocks that require `ed25519-dalek`, because they included a transaction that verified differently as when using `ed25519-zebra`.

This feature can be enabled in the following way. In the chain service file, directly after the client is created, the following code should be added:

```
use sc_client_api::ExecutorProvider;
client.execution_extensions().set_extensions_factory(
	sc_client_api::execution_extensions::ExtensionBeforeBlock::<Block, sp_io::UseDalekExt>::new(BLOCK_NUMBER_UNTIL_DALEK_SHOULD_BE_USED)
);
```

### CODE REVIEW

**Comment by davxy** [comment]
Excluded my concern reported in the comment, the PR LGTM.

File: primitives/io/src/lib.rs [hunk 1]
```diff
@@ -2010,4 +2057,20 @@ mod tests {
...
+		ext.register_extension(UseDalekExt::default());
+
+		// With dalek the zero signature should fail to verify.
+		ext.execute_with(|| {
+			assert!(!crypto::ed25519_verify(&zero_ed_sig(), &Vec::new(), &zero_ed_pub()));
+		});
+
+		// But with zebra it should work.
+		BasicExternalities::default().execute_with(|| {
+			assert!(crypto::ed25519_verify(&zero_ed_sig(), &Vec::new(), &zero_ed_pub()));
```
[the remark this answers is not in the document]
Comment (davxy) [reply primitives/io/src/lib.rs:2073]: Just for future reference:
- https://github.com/ZcashFoundation/ed25519-zebra/issues/64
- https://github.com/jedisct1/libsodium/issues/112

**Comment by librelois** [comment]
@bkchr Is it possible to merge this PR soon? We still need it at Moonbeam to be able to sync the old blocks while being compatible with the relay.

**Comment by bkchr** [comment]
> @bkchr Is it possible to merge this PR soon? We still need it at Moonbeam to be able to sync the old blocks while being compatible with the relay.

Yeah, I will merge now after CI is green. I was distracted by other stuff :P

**Comment by bkchr** [comment]
bot merge

**Comment by bkchr** [comment]
bot merge

**Comment by dzmitry-lahoda** [comment]
Is it the case:

1. My substrate parachain on code before https://github.com/paritytech/substrate/commit/cf36d2f7eaaee27516c601141a2ea0acd3db641c 
2. If I update to https://github.com/paritytech/substrate/commit/cf36d2f7eaaee27516c601141a2ea0acd3db641c I potentially will break my chain.
3. If I upgrade to https://github.com/PureStake/moonbeam/pull/1982/files , but not apply extension, I may have issues
4. So really I should use extensions https://github.com/PureStake/moonbeam/pull/1982/files
5. In extension I should set block in future so we upgrade runtime and it will use zebra in some future.
6. If we do not use extension, but use new code, we will run zebra on old blocks, which is bad

right?

**Comment by bkchr** [comment]
> 6\. If we do not use extension, but use new code, we will run zebra on old blocks, which is bad

It could may lead to issues importing old blocks.

> 5\. In extension I should set block in future so we upgrade runtime and it will use zebra in some future.

Yes. This will be like a hard fork.

**Comment by aliXsed** [comment]
@bkchr Why the dalek end block number is not set for polkadot parachains inside cumulus yet? (I checked the release 0.9.37)

**Comment by bkchr** [comment]
> @bkchr Why the dalek end block number is not set for polkadot parachains inside cumulus yet? (I checked the release 0.9.37)

You only need this extension when there was before (in some block in the past) some signature that was validated differently in dalek vs zebra.

**Pull request merged**